features/lwipstack/LWIPInterface.cpp

```diff
@@ -212,19 +213,22 @@ void LWIP::Interface::netif_status_irq(struct netif *netif)
             dns_addr_has_to_be_added = true;
         }
 #endif
-       if (dns_addr_has_to_be_added && !interface->blocking) {
+        if (dns_addr_has_to_be_added && !interface->blocking) {
             add_dns_addr(&interface->netif);
         }
 
-
         if (interface->has_addr_state & HAS_ANY_ADDR) {
             interface->connected = NSAPI_STATUS_GLOBAL_UP;
         }
-    } else {
+    } else if (!netif_is_up(&interface->netif) && netif_is_link_up(&interface->netif)) {
+        // disconnect() was called, because netif_is_link_up is returning true even though netif is down.
```

Comment:
After discussing about this, I'm still wondering that there should be one more test.
What happens if you disconnect the ethernet cable, and then call `interface->disconnect()`

I would assume that in that case both functions are returning `false` in which case the generated event is probably not correct.
@michalpasztamobica  Can you verify?

Reply:
Excellent point, you are right - if I call `interface->disconnect()` after I unplugged the cable there is no status change (it should change to `DISCONNECTED`.
I will try to improve the condition, but I am afraid we need to entangle the connection status and create kind of a simple state machine, which considers previous state in its logic.